web/webkit/src/main/scala/net/liftweb/http/HtmlNormalizer.scala

```diff
@@ -0,0 +1,291 @@
+package net.liftweb
+package http
+
+import scala.xml._
+import scala.collection.immutable.Vector
+
+import js.JsCmd
+import js.JsCmds.Noop
+import js.JE.{AnonFunc, Call, JsRaw}
+
+import util.Helpers._
+
+/**
+ * Represents an HTML attribute for an event handler. Carries the event name and
+ * the JS that should run when that event is triggered as a String.
+ */
+private case class EventAttribute(eventName: String, jsString: String)
+private object EventAttribute {
+  /**
+   * Some elements allow a URL attribute to take a `javascript:(//)`-style
+   * URL instead of setting an `on*` event in order to invoke JS. For example,
+   * you can (and Lift does) set a form's `action` attribute to
+   * `javascript://(some JS)` instead of setting `onsubmit` to `(someJS);
+   * return false`.
+   *
+   * This is a map from those attribute names to the corresponding JS event
+   * that would be used to execute that JS when it isn't run in line.
+   */
+  val eventsByAttributeName =
+    Map(
+      "action" -> "submit",
+      "href" -> "click"
+    )
+
+  object EventForAttribute {
+    def unapply(attributeName: String): Option[String] = {
+      eventsByAttributeName.get(attributeName)
+    }
+  }
+}
+
+/**
+ * Helper class that performs certain Lift-specific normalizations of HTML
+ * represented as `NodeSeq`s:
+ *  - Fixes URLs for `link`, `a`, `form`, and `script` elements to properly
+ *    prepend the container's context path in case these are absolute URLs.
+ *  - Extracts event attributes (replacing them with
+ *    `[[LiftRules.attributeForRemovedEventAttributes]]` if needed), returning
+ *    instead JavaScript that will attach the corresponding event handler to
+ *    that element.
+ *  - Provides for caller-specific additional processing for each node.
+ */
+private[http] final object HtmlNormalizer {
+  // Fix URLs using Req.normalizeHref and extract JS event attributes for
+  // putting into page JS. Returns a triple of:
+  //  - The optional id that was found in this set of attributes.
+  //  - The normalized metadata.
+  //  - A list of extracted `EventAttribute`s.
+  private def normalizeUrlAndExtractEvents(
+    attributeToNormalize: String,
+    attributes: MetaData,
+    contextPath: String,
+    shouldRewriteUrl: Boolean, // whether to apply URLRewrite.rewriteFunc
+    eventAttributes: List[EventAttribute] = Nil
+  ): (Option[String], MetaData, List[EventAttribute]) = {
```

Comment:
It could be. I vacillated a bit on that front but basically couldn't come up with a proper name for the things it was doing. That's because it's currently doing 3 semi-independent things (extracting id, normalizing URLs, and extracting event attributes). On its face, this seems like a bad thing (separation of concerns and such), but because `MetaData` is a linked list, the quickest way to do this is recursively starting at the beginning—and because this is run as part of the core HTML rendering pipeline I didn't want to get too crazy on potentially sacrificing performance for clarity.

Anyway, since the method itself is private and is only used in one place (where the sub-parts are immediately unpacked on assignment), I settled on a tuple + a comment.